File: AccuPay/Domain/Calendars/CalendarDay.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("calendarday")>
    Public Class CalendarDay
        Implements IPayrate

        <Key>
        Public Property RowID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property Updated As Date?

        Public Property UpdatedBy As Integer?

        Public Property CalendarID As Integer?

        Public Property DayTypeID As Integer?

        Public Property [Date] As Date

        Public Property Description As String

        Public Overridable Property DayType As DayType

        Public ReadOnly Property RegularRate As Decimal Implements IPayrate.RegularRate
            Get
                Return DayType.RegularRate
            End Get
        End Property

        Public ReadOnly Property OvertimeRate As Decimal Implements IPayrate.OvertimeRate
            Get
                Return DayType.OvertimeRate
            End Get
        End Property

        Public ReadOnly Property NightDiffRate As Decimal Implements IPayrate.NightDiffRate
            Get
                Return DayType.NightDiffRate
            End Get
        End Property

        Public ReadOnly Property NightDiffOTRate As Decimal Implements IPayrate.NightDiffOTRate
            Get
                Return DayType.NightDiffOTRate
            End Get
        End Property

        Public ReadOnly Property RestDayRate As Decimal Implements IPayrate.RestDayRate
            Get
                Return DayType.RestDayRate
            End Get
        End Property

        Public ReadOnly Property RestDayOTRate As Decimal Implements IPayrate.RestDayOTRate
            Get
                Return DayType.RestDayOTRate
            End Get
        End Property

        Public ReadOnly Property RestDayNDRate As Decimal Implements IPayrate.RestDayNDRate
            Get
                Return DayType.RestDayNDRate
            End Get
        End Property

        Public ReadOnly Property RestDayNDOTRate As Decimal Implements IPayrate.RestDayNDOTRate
            Get
                Return DayType.RestDayNDOTRate
            End Get
        End Property

        Public ReadOnly Property IsRegularDay As Boolean Implements IPayrate.IsRegularDay
            Get
                Return DayType.Name = "Regular Day"
            End Get
        End Property

        Public ReadOnly Property IsRegularHoliday As Boolean Implements IPayrate.IsRegularHoliday
            Get
                Return DayType.Name = "Regular Holiday"
            End Get
        End Property

        Public ReadOnly Property IsSpecialNonWorkingHoliday As Boolean Implements IPayrate.IsSpecialNonWorkingHoliday
            Get
                Return DayType.Name = "Special Non-Working Holiday"
            End Get
        End Property

        Public ReadOnly Property IsHoliday As Boolean Implements IPayrate.IsHoliday
            Get
                Return IsRegularHoliday OrElse IsSpecialNonWorkingHoliday
            End Get
        End Property

        Private ReadOnly Property IPayrate_Date As Date Implements IPayrate.Date
            Get
                Return [Date]
            End Get
        End Property

    End Class

End Namespace

File: Accupay.Data/Entities/CalendarDay.cs
using AccuPay.Data.Helpers;
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("calendarday")]
    public class CalendarDay : IPayrate
    {
        [Key]
        public int? RowID { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        public int? CreatedBy { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime? Updated { get; set; }

        public int? UpdatedBy { get; set; }

        public int? CalendarID { get; set; }

        public int? DayTypeID { get; set; }

        public DateTime Date { get; set; }

        public string Description { get; set; }

        public virtual DayType DayType { get; set; }

        public decimal RegularRate => DayType.RegularRate;

        public decimal OvertimeRate => DayType.OvertimeRate;

        public decimal NightDiffRate => DayType.NightDiffRate;

        public decimal NightDiffOTRate => DayType.NightDiffOTRate;

        public decimal RestDayRate => DayType.RestDayRate;

        public decimal RestDayOTRate => DayType.RestDayOTRate;

        public decimal RestDayNDRate => DayType.RestDayNDRate;

        public decimal RestDayNDOTRate => DayType.RestDayNDOTRate;

        public bool IsRegularDay => DayType?.DayConsideredAs == CalendarConstant.RegularDay || DayType is null;

        public bool IsSpecialNonWorkingHoliday => DayType?.DayConsideredAs == CalendarConstant.SpecialNonWorkingHoliday;

        public bool IsRegularHoliday => DayType?.DayConsideredAs == CalendarConstant.RegularHoliday || DayType?.DayConsideredAs == CalendarConstant.DoubleHoliday;

        public bool IsDoubleHoliday => DayType?.DayConsideredAs == CalendarConstant.DoubleHoliday;

        public bool IsHoliday => IsRegularHoliday || IsSpecialNonWorkingHoliday;
    }
}
